### ISSUE TITLE
Parameter `metric` not (properly) implemented in `makeStackedLearner`

### ISSUE THREAD

**Issue by philippstats** [mlr-org/mlr#808]
Presently `method = "hill.climb" `of function makeStackedLearner does not support metrics like they are already available in mlr.

`metric` is kind of hardcoded and only can take two arguments, which are prediction and truth. However the measure functions for supervised learning in mlr need the arguments `task`, `model`, and `pred`.

For the the simple case of `init = 0`, metrics can be implemented the following way (if I'm right):
(Code is based on pull-request #807; new lines is tagged with '# new')

```r
hillclimbBaseLearners = function(learner, task, replace = TRUE, init = 5, bagprob = 0.5, bagtime = 20,
  metric = NULL, ...) {

  assertFlag(replace)
  assertInt(init, lower = 0, upper = length(learner$base.learners)) #807
  assertNumber(bagprob, lower = 0, upper = 1)
  assertInt(bagtime, lower = 1)

  td = getTaskDescription(task)
  type = ifelse(td$type == "regr", "regr",
                ifelse(length(td$class.levels) == 2L, "classif", "multiclassif"))
  if (is.null(metric)) {
    if (type == "regr") {
      metric = function(pred, true) mean((pred-true)^2)
    } else {
      metric = function(pred, true) {
        pred = colnames(pred)[max.col(pred)]
        tb = table(pred, true)
        return( 1- sum(diag(tb))/sum(tb) )
      }
    }
  } else { # new
    assertClass(metric, "Measure")  # new
    metric = metric$fun # new # new
  }

  bls = learner$base.learners
  if (type != "regr") {
    for (i in 1:length(bls)) {
      if (bls[[i]]$predict.type == "response")
        stop("Hill climbing algorithm only takes probability predict type for classification.")
    }
  }
  # cross-validate all base learners and get a prob vector for the whole dataset for each learner
  base.models = resres = probs = vector("list", length(bls)) # new
  rin = makeResampleInstance(learner$resampling, task = task)
  for (i in seq_along(bls)) {
    bl = bls[[i]]
    resres[[i]] = r = resample(bl, task, rin, show.info = FALSE) # new
    if (type == "regr") {
      probs[[i]] = matrix(getResponse(r$pred), ncol = 1)
    } else {
      probs[[i]] = getResponse(r$pred, full.matrix = TRUE)
      colnames(probs[[i]]) = task$task.desc$class.levels
    }
    # also fit all base models again on the complete original data set
    base.models[[i]] = train(bl, task)
  }
  names(probs) = names(bls)
  names(resres) = names(bls) # new

  # add true target column IN CORRECT ORDER
  tn = getTaskTargetNames(task)
  test.inds = unlist(rin$test.inds)

  # now start the hill climbing
  probs = lapply(probs, function(x) x[order(test.inds), , drop = FALSE])
  probs[[tn]] = getTaskTargets(task)[test.inds]
  probs[[tn]] = probs[[tn]][order(test.inds)]
  # probs = probs[order(test.inds), , drop = FALSE]
  m = length(bls)
  weights = rep(0, m)
  flag = TRUE
  for (bagind in 1:bagtime) {
    # bagging of models
    bagsize = ceiling(m*bagprob)
    bagmodel = sample(1:m, bagsize)
    bagweight = rep(0, m) #807

    # Initial selection of strongest learners
    inds = NULL
    if (init>0) {
      score = rep(Inf, m)
      for (i in bagmodel) {
        score[i] = metric(probs[[i]], probs[[tn]])
      }
      inds = order(score)[1:init]
      bagweight[inds] = 1 #807
    }

    selection.size = init
    selection.ind = inds
    # current.prob = rep(0, nrow(probs))
    current.prob = matrix(0, nrow(probs[[1]]), ncol(probs[[1]]))
    old.score = Inf
    if (selection.size>0) {
      current.prob = Reduce('+', probs[selection.ind])
      old.score = metric(current.prob/selection.size, probs[[tn]])
    }
    flag = TRUE

    while (flag) {
      score = rep(Inf, bagsize)
      for (i in bagmodel) {
        #score[i] = metric( (probs[[i]]+current.prob)/(selection.size+1), probs[[tn]] ) # new
        score[i] = metric(task, model = base.models[[i]], pred = resres[[i]]$pred) # new
      }
      inds = order(score)
      if (!replace) {
        ind = setdiff(inds, selection.ind)[1]
      } else {
        ind = inds[1]
      }

      new.score = score[ind]
      if (old.score-new.score<1e-8) {
        flag = FALSE
      } else {
        current.prob = current.prob+probs[[ind]]
        weights[ind] = weights[ind]+1
        selection.ind = c(selection.ind, ind)
        selection.size = selection.size+1
        old.score = new.score
      }
    }
    weights = weights + bagweight #807
  }
  weights = weights/sum(weights)

  list(method = "hill.climb", base.models = base.models, super.model = NULL,
       pred.train = probs, weights = weights)
}

```
For `init > 0`/`selection.size > 0` it is more complicated. @berndbischl @giuseppec Can we talk about that?

Run stacking:

```
bls = list(
  makeLearner("classif.ksvm", predict.type = "prob"),
  makeLearner("classif.randomForest", predict.type = "prob"),
  makeLearner("classif.kknn", id = "bls.knn2", predict.type = "prob", k = 2),
  makeLearner("classif.kknn", id = "bls.knn3", predict.type = "prob", k = 3),
  makeLearner("classif.kknn", id = "bls.knn4", predict.type = "prob", k = 4),
  makeLearner("classif.kknn", id = "bls.knn5", predict.type = "prob", k = 5)
)

# True positive
stk = makeStackedLearner(bls, method = "hill.climb", predict.type = "response",
  parset = list(init = 0, bagprob = 0.5, bagtime = 5, metric = tp))
res = resample(stk, task = pid.task, resampling = hout)

# MMCE
stk = makeStackedLearner(bls, method = "hill.climb", predict.type = "response",
  parset = list(init = 0, bagprob = 0.5, bagtime = 5, metric = mmce))
res = resample(stk, task = pid.task, resampling = cv2)
```

We "just" need to extract the resampling result to have access to `r$pred` (e.g. for "True positive" i.e. `r$pred$data$truth`, `r$pred$data$response`). `task` (=`task`) and `model` (= `base.models[[i]]`) are already available.

**Comment by berndbischl** [comment]
You said in another thread you need some kind of help here. then please ask.

**Comment by philippstats** [comment]
Implemented in #1041
